modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -0,0 +1,560 @@
+
+    private final Long2ObjectMap<PartitionPruningColumns> result = new Long2ObjectArrayMap<>();
+
+    /**
+     * Extracts partition pruning metadata from the given physical plan.
+     * This method traverses a physical plan and attempts to extract metadata from operators that support it
+     * and includes such metadata is that complete (e.g. values of all colocation keys present in a scan predicate).
+     *
+     * @param rel Physical plan.
+     * @return Partition pruning metadata.
+     */
+    public PartitionPruningMetadata go(IgniteRel rel) {
+        rel.accept(this);
+
+        return new PartitionPruningMetadata(result);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public IgniteRel visit(IgniteIndexScan rel) {
+        RexNode condition = rel.condition();
+
+        IgniteTable table = rel.getTable().unwrap(IgniteTable.class);
+        assert table != null : "No table";
+
+        extractFromTable(rel.sourceId(), table, condition, rel.getCluster().getRexBuilder());
+
+        return super.visit(rel);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public IgniteRel visit(IgniteTableScan rel) {
+        RexNode condition = rel.condition();
+
+        IgniteTable table = rel.getTable().unwrap(IgniteTable.class);
+        assert table != null : "No table";
+
+        extractFromTable(rel.sourceId(), table, condition, rel.getCluster().getRexBuilder());
+
+        return rel;
+    }
+
+    private void extractFromTable(long sourceId, IgniteTable table, @Nullable RexNode condition, RexBuilder rexBuilder) {
+        if (condition == null) {
+            return;
+        }
+
+        IgniteDistribution distribution = table.distribution();
+        if (!distribution.function().affinity()) {
+            return;
+        }
+
+        PartitionPruningColumns metadata = extractMetadata(distribution.getKeys(), condition, rexBuilder);
+
+        if (metadata != null) {
+            result.put(sourceId, metadata);
+        }
+    }
+
+    /** Extracts values of colocated columns from the given condition. */
+    @VisibleForTesting
+    @Nullable
+    public static PartitionPruningColumns extractMetadata(List<Integer> keys, RexNode condition, RexBuilder rexBuilder) {
+        Result res = extractMetadata(condition, keys, rexBuilder, false);
+
+        // Both unknown condition and additional condition can not be used to extract metadata.
+        if (res == Result.UNKNOWN || res == Result.RESTRICT) {
+            return null;
+        }
+
+        PruningColumnSets columnSets;
+
+        if (res instanceof PruningColumnSet) {
+            PruningColumnSet columnSet = (PruningColumnSet) res;
+            columnSets = new PruningColumnSets();
+            columnSets.candidates.add(columnSet);
+        } else {
+            columnSets = (PruningColumnSets) res;
+        }
+
+        for (PruningColumnSet columnSet : columnSets.candidates) {
+            if (!columnSet.columns.keySet().containsAll(keys)) {
```

Comment:
Seems, `containsAll` uses boxed collections, because `keys` kind is of Calcite `ImmutableIntList`.

Reply:
Fixed automatically by using `IntList` instead of a boxed `java.util.Lis`t because `IntSet` has a method `containAll(IntCollection)`.